clang/lib/AST/ExprConstant.cpp

```diff
@@ -14951,6 +14961,36 @@ bool FloatExprEvaluator::VisitCallExpr(const CallExpr *E) {
   default:
     return false;
 
+  case Builtin::BIfrexp:
+  case Builtin::BIfrexpf:
+  case Builtin::BIfrexpl:
+  case Builtin::BI__builtin_frexp:
+  case Builtin::BI__builtin_frexpf:
+  case Builtin::BI__builtin_frexpl:
+  case Builtin::BI__builtin_frexpf16:
+  case Builtin::BI__builtin_frexpf128: {
+    const auto *FDecl = E->getDirectCallee();
+    clang::LangStandard::Kind ConstExprSinceVersion =
+        FDecl->getConstexprBuiltinSinceVersion();
+    if (ConstExprSinceVersion > Info.Ctx.getLangOpts().LangStd)
+      return false;
+    LValue Pointer;
+    if (!EvaluateFloat(E->getArg(0), Result, Info) ||
+        !EvaluatePointer(E->getArg(1), Pointer, Info))
```

Comment:
I think the builtin should be unconditionally constexpr in Clang. 

And we should certainly not store additional info for  that in `FunctionDecl`, which has a huge cost in terms of AST size.

Instead, the builtin should become constexpr, and libc++  can decide not to make the wrapper constexpr in older language modes, or to have some availability warning in older language modes.

Reply:
@hubert-reinterpretcast I think if we wanted to do something about the non-prefixed builtin, I would prefer
 - that we table-gen the information
 - that the version is encoded as an attribute (it might just be an implementation detail attribute) - so as to not bloat the AST
 - that we make it a diagnostic that users can turn into an error or not (and then we can discuss what the default should be)